_posts/2020-01-09-rxjs-wstep.md

```diff
@@ -0,0 +1,180 @@
+### Strumień zimny
+Strumień zimny nie będzie emitować (produkować) danych aż do momentu, gdy ktoś (obserwator) zacznie go obserwować. Wyemituje odrębną wartość dla każdego nowego obserwatora - te wartości nie są współdzielone. Przykładowo: wysłanie żądania GET do serwera.
+```typescript
+this.httpClient.get<ServerResponse>('someUrl')
+```
+### Strumień ciepły
+Strumień ciepły, w przeciwieństwie do zimnego, emituje dane niezależnie od tego, czy ktoś na niego nasłuchuje. Każdy obserwator operuje na współdzielonym zasobie danych - dwóch obserwatorów otrzyma tą samą wartość w momencie wyemitowania danej przez strumień. Przykładowo: wydarzenia 'click'
+```typescript
+import {fromEvent} from 'rxjs';
+ 
+fromEvent(document, 'click')
+```
+
+## Observable, wzorzec Obserwator
+Mając już wiedzę czym jest strumień i z jakimi rodzajami strumieni możemy się spotkać możemy przejść do opisania podstawowego konceptu RxJS: Observable.
+Observable jest obiektem reprezentującym strumień danych. Implementuje wzorzec projektowy _Obserwator_, który zakłada istnienie bytu przechowującego listę obiektów - obserwatorów, nasłuchujących na jakiekolwiek zmiany stanu danego bytu, powiadamiającego każdego z nich o zmianie poprzez wywołanie funkcji przez nich przekazanych (callback).
+
+Najprostszego Observable możemy stworzyć za pomocą funkcji statycznej _of_.
+```typescript
+// obiekt Observable, emitujący wartości liczbowe od 1 do 5
+const numbers$: Observable<number> = of(1,2,3,4,5);
+
+```
+Obiekt numbers$ jest definicją strumienia danych typu _number_. Jest to tylko i wyłącznie szablon strumienia. W tym przypadku stworzyliśmy strumień zimny. Znamy zbiór danych (1,2,3,4,5), jednak dane zaczną być emitowane dopiero w momencie rozpoczęcia nasłuchiwania na dany strumień przez obserwatora. Pod strumień 'podłączamy' się za pomocą funkcji subscribe().
+### subscribe() i unsubscribe()
+Funkcja subscribe jako parametr oczekuje obiektu, który definiuje trzy funkcje: _next_, _error_ oraz _complete_.
+```typescript
+const subscription = numbers$.subscribe({
+    next(value) {},
+    error(err) {},
+    complete() {}
+});
+```
+Każda z tych funkcji jest _callbackiem_, który jest wywoływany w poszczególnych momentach przepływu danych przez strumień. Funkcja _next(value)_ wywoływana jest za każdym razem, gdy strumień emituje pojedyńczą wartość - czyli w naszym przypadku funkcja _next()_ zostanie wywołana 5 razy, raz dla każdej z cyfr z zakresu 1-5. Callback _error(err)_ zostanie wywołany w momencie, gdy strumień zostanie nienaturalnie zamknięty lub przerwany. _Complete_ jest ostatnim callbackiem, który wywołoywany jest po zamknięciu strumienia.
```

Comment:
pojedyńczą ==> pojedynczą

zostanie wywołana 5 razy, raz dla każdej z cyfr z zakresu 1-5 ==> zamiast _raz_ powinno być _po razie_, po to ma inne znaczenie, a jednocześnie poleciłbym spróbować by uniknąć powtórzenia przez taki zabieg:
zostanie wywołana **5-krotnie**, **po razie** dla każdej z cyfr z zakresu 1-5